File: domain/src/main/java/io/prizy/domain/contest/service/ContestService.java
package io.prizy.domain.contest.service;

import java.time.Instant;
import java.util.Collection;
import java.util.Optional;
import java.util.UUID;

import io.prizy.domain.contest.model.Contest;
import io.prizy.domain.contest.model.Pack;
import io.prizy.domain.contest.ports.ContestRepository;
import io.prizy.domain.contest.ports.ContestTemplateCompiler;
import io.prizy.domain.contest.ports.PackRepository;
import io.prizy.toolbox.useragent.UserAgent;
import io.prizy.toolbox.useragent.UserAgentParser;
import lombok.RequiredArgsConstructor;
import org.apache.commons.lang3.RandomStringUtils;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;


@Service
@Transactional
@RequiredArgsConstructor
public class ContestService {

  private static final Integer ACCESS_CODE_LENGTH = 6;
  private static final String ACCESS_CODE_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

  private final ContestRepository contestRepository;
  private final PackRepository packRepository;
  private final ContestTemplateCompiler templateCompiler;

  public Optional<Contest> byId(UUID id) {
    return contestRepository.byId(id);
  }

  public Optional<Contest> byPackId(UUID packId) {
    return contestRepository.byPackId(packId);
  }

  public Collection<Contest> byIds(Collection<UUID> ids) {
    return contestRepository.byIds(ids);
  }

  public Collection<Contest> contests() {
    return contestRepository.list();
  }

  public Collection<Contest> activeContests() {
    return contestRepository.startedBeforeAndEndingAfter(Instant.now());
  }

  public Collection<Contest> listAllContests() {
    return contestRepository.list();
  }

  public String generateRandomAccessCode() {
    return RandomStringUtils.random(ACCESS_CODE_LENGTH, ACCESS_CODE_ALPHABET);
  }

  public Optional<Pack> packById(UUID id) {
    return packRepository.byId(id);
  }

  public String contestRules(UUID contestId, String userAgent) {
    var contest = contestRepository
      .byId(contestId)
      .orElseThrow(() -> new IllegalArgumentException("Contest not found"));
    var platform = UserAgentParser.parse(userAgent).map(UserAgent::platform).orElse("Unknown");
    return templateCompiler.rulesTemplate(contest, platform);
  }

}

File: domain/src/main/kotlin/io/prizy/domain/contest/service/ContestService.kt
package io.prizy.domain.contest.service

import io.prizy.domain.contest.command.CreateContest
import io.prizy.domain.contest.command.CreatePack
import io.prizy.domain.contest.command.UpdateContest
import io.prizy.domain.contest.event.ContestCreated
import io.prizy.domain.contest.event.ContestUpdated
import io.prizy.domain.contest.model.Contest
import io.prizy.domain.contest.model.Pack
import io.prizy.domain.contest.ports.ContestPublisher
import io.prizy.domain.contest.ports.ContestRepository
import io.prizy.domain.contest.ports.ContestTemplateCompiler
import io.prizy.domain.contest.ports.PackRepository
import io.prizy.toolbox.coroutines.runDetached
import java.util.UUID

/**
 *  @author Nidhal Dogga
 *  @created 4/24/2022 12:15 AM
 */

class ContestService(
  private val packRepository: PackRepository,
  private val templateCompiler: ContestTemplateCompiler,
  private val contestPublisher: ContestPublisher,
  private val contestRepository: ContestRepository,
) {

  suspend fun contestById(id: UUID): Contest? {
    return contestRepository.byId(id)
  }

  suspend fun contests(): Collection<Contest> {
    return contestRepository.list()
  }

  suspend fun userContests(userId: UUID): Collection<Contest> {
    return contestRepository.list()
  }

  suspend fun listAllContests(): Collection<Contest> {
    return contestRepository.list()
  }

  suspend fun createContest(command: CreateContest): Contest = contestRepository
    .create(command.contest)
    .also { contest -> runDetached { contestPublisher.publish(ContestCreated(contest)) } }

  suspend fun updateContest(command: UpdateContest): Contest = contestRepository
    .update(command.contest)
    .also { contest -> runDetached { contestPublisher.publish(ContestUpdated(contest)) } }

  suspend fun packById(id: UUID): Pack? {
    return packRepository.byId(id)
  }

  suspend fun contestRules(contestId: UUID): String = contestRepository
    .byId(contestId)
    ?.let { contest -> templateCompiler.rulesTemplate(contest) }
    ?: throw IllegalStateException("Contest[$contestId] not found")

  private val CreateContest.contest: Contest
    get() = Contest(
      id = UUID.randomUUID(),
      name = name,
      description = description,
      fromDate = fromDate,
      toDate = toDate,
      adultOnly = adultOnly,
      category = category,
      coverMediaId = coverMediaId,
      mediaIds = mediaIds,
      cost = cost,
      facebookPage = facebookPage,
      instagramPage = instagramPage,
      website = website,
      newsletterSubscription = newsletterSubscription,
      merchant = merchant,
      boardId = boardId,
      packs = packs.map { command -> command.pack }
    )

  private val UpdateContest.contest: Contest
    get() = Contest(
      id = id,
      name = name,
      description = description,
      fromDate = fromDate,
      toDate = toDate,
      adultOnly = adultOnly,
      category = category,
      coverMediaId = coverMediaId,
      mediaIds = mediaIds,
      cost = cost,
      facebookPage = facebookPage,
      instagramPage = instagramPage,
      website = website,
      newsletterSubscription = newsletterSubscription,
      merchant = merchant,
      boardId = boardId,
      packs = packs.map { command -> command.pack }
    )

  private val CreatePack.pack: Pack
    get() = when (this) {
      is CreatePack.Product -> Pack.Product(
        id = UUID.randomUUID(),
        name = name,
        lastWinnerPosition = lastWinnerPosition,
        firstWinnerPosition = firstWinnerPosition,
        value = value,
        mediaId = mediaId,
        quantity = quantity,
      )
      is CreatePack.Coupon -> Pack.Coupon(
        id = UUID.randomUUID(),
        name = name,
        lastWinnerPosition = lastWinnerPosition,
        firstWinnerPosition = firstWinnerPosition,
        code = code,
        expiration = expiration,
      )
    }
}
